app.py

```diff
@@ -107,47 +118,74 @@ def login():
     flask.session['macaroon_root'] = root
 
     return oid.try_login(
-        UBUNTU_SSO_URL,
+        'https://login.ubuntu.com',
         ask_for=['email', 'nickname'],
         ask_for_optional=['fullname'],
         extensions=[openid_macaroon]
     )
 
 
 @oid.after_login
-def create_or_login(resp):
+def after_login(resp):
     flask.session['openid'] = resp.identity_url
-
     flask.session['macaroon_discharge'] = resp.extensions['macaroon'].discharge
 
+    return flask.redirect(oid.get_next_url())
+
+
+@app.route('/app-detail')
+def get_app_detail():
+    if not is_authenticated():
+        return redirect_to_login()
+
     authorization = get_authorization_header(
         flask.session['macaroon_root'],
         flask.session['macaroon_discharge']
     )
+
     headers = {
         'X-Ubuntu-Series': '16',
         'X-Ubuntu-Architecture': 'amd64',
         'Authorization': authorization
     }
     url = (
-        'https://api.snapcraft.io/api/v1/snaps/details'
-        '/documentation-builder?revision=3'
+        "https://api.snapcraft.io/api/v1/snaps/details/"
+        "documentation-builder?revision=3"
     )
 
     response = requests.request(url=url, method='GET', headers=headers)
-    response.raise_for_status()
+
+    if response.status_code > 400:
+        verified = verify_macaroon(
+            flask.session['macaroon_root'],
+            flask.session['macaroon_discharge']
+        )
+
+        # Macaroon not valid anymore, needs refresh
+        if verified['account'] is None:
+            empty_session()
+            return flask.redirect('/login')
+
+        # Not authorized content
+        if response.status == 401 and not verified['allowed']:
+            return '<h1>Not allowed</h1>'
```

Comment:
Rather than implementing our own error behaviours here (and returning 200 statuses) I'd rather do `response.raise_for_status()` in these situations (which will throw a python error for now) and then eventually, but not now, pass it through to native Flask error handing behaviour